### ISSUE TITLE
Crypto and Security Audit

### ISSUE THREAD

**Issue by danpalmer** [irungentoo/toxcore#121]
Would be good to get this audited by an external person/group/company so the public can have a bit more faith in the security of the application, especially after the recent Cryptocat failures stemming from the developers having a lack of understanding of cryptographic principles.

**Comment by iamunknown2** [comment]
@irungentoo, could you please not assume that secure library = secure protocol? It's about as logical as strong material = strong building. It isn't just about the libraries you use; it's also about **how** you use it. For example, no matter how strong the material I make the cards in my house of cards out of, it's still going to crash to the ground when I kick it.

Criticism aside, Tox **really** needs an audit to check if Tox is secure or not.

**Comment by GrayHatter** [comment]
@iamunknown2 yes, we're aware and it's on our The problem is a security audit that's actually worth it, is very expensive. So until we get the money, it's pending.

**Comment by iamunknown2** [comment]
> I'm not going to comment on your strawman fallacy. But I want everyone else reading to be aware of it.

> I suggest those who think Tox isn't secure to read:
> http://nacl.cr.yp.to/
>
> This is more specifically what Tox uses for crypto:
> http://nacl.cr.yp.to/box.html
>
> Then after that if you still have doubts you can open the code in a text editor and try to find issues with it.
>
> Tox is secure.
>
> A real crypto audit would however be very welcome.

He's implying that Tox is secure is because of the fact that it uses NaCl.

**Comment by GrayHatter** [comment]
It's a strawman when you ignore the context. (Talking about MITM risks.) Or the follow up 

>I probably should have said "The design of Tox has a very high probability of being secure" instead but I find saying "Tox is secure" is more to the point.

Also, if there was a mistake in the implementation of nacl, it'd be glaring enough for anyone to find. Else, you're saying the crypto box API from nacl is in secure.

**Comment by GrayHatter** [comment]
Right, and because of it's use of the NaCl api. If NaCl is secure (it is), and we're using the API correctly. Tox is secure by proxy. This is speaking to the encryption used. An audit of the code may find bugs but the crypto would still be secure. (Assuming we're using the API correctly.)

**Comment by konklone** [comment]
> Right, and because of it's use of the NaCl api. If NaCl is secure (it is), and we're using the API correctly. Tox is secure by proxy. This is speaking to the encryption used. An audit of the code may find bugs but the crypto would still be secure. (Assuming we're using the API correctly.)

Crypto can be rendered insecure when used insecurely. How you use their API matters, what version of NaCl you're using matters (and your processes for staying up to date).

For example, I can use bcrypt in my web app to manage password storage (which I have done before). That's not enough to say that my passwords are stored securely. It matters what methods I call, and what values I pass into those methods. It matters what other operations I might be performing on the data that bcrypt encrypted. It matters whether some other part of my code compromises the RNG, or monkeypatched in some strange behavior that might accidentally effect bcrypt's operation.

I can say "I use bcrypt" and that should provide some general comfort to folks asking about our choice of methods. But from the vantage point of a *professional security analysis*, that's completely insufficient. Just saying "we use NaCl" is fine as a talking point, but not enough to defend your app's security posture to professionals.

**Comment by nurupo** [comment]
Yep, totally agree that using crypto library doesn't mean that an app is secure. Even though the NaCl/libsodium might have quite friendly APIs that makes it hard to shoot yourself in the foot, it's still possible to do so. For example, simply forgetting to change the nonce, which NaCl doesn't do for you, could render all of the security useless. So the way you use the crypto library is important.

**Comment by bill-auger** [comment]
@irungentoo  commented on Feb 21, 2014
> A real crypto audit would however be very welcome.

@GrayHatter commented on Mar 1, 2016
> The problem is a security audit that's actually worth it, is very expensive. So until we get the money, it's pending.

someone on IRC showed me this today and i thought immediately of this issue - it appears that mozilla is sponsoring security audits for security-minded FOSS projects - i think tox is a prime candidate for this program and most likely meets the project requirements

the announcement is here
https://blog.mozilla.org/blog/2016/06/09/help-make-open-source-secure/

the project requirements are here
https://wiki.mozilla.org/MOSS/Secure_Open_Source

the application is here
https://docs.google.com/forms/d/1f0xSg9XM8v7YGdZ_FzeE67ggckbAsg6sH1mpQ4buTQE/viewform

the application states that anyone can nominate the project for consideration - they need not be associated with the project

**Comment by GrayHatter** [comment]
You should consider nominating tox then.

@JFreegman and I were talking about this exact thing on IRC today as well, we came to the agreement that tox really isn't ready for an audit. With three large pull requests in the pipe (TCPDHT, Groupchats, Multi-device).

That's not to say we wouldn't welcome it. Merely that we won't be nominating toxcore for an audit yet.

Can't speak for JFreegman, but I'm often wrong, so this shouldn't discourage you if you still think tox is perfect for this program.
